### PR TITLE
Feature : Checkbox Group widget

### LINKED ISSUE #3344: [Feature] Checkbox Group Widget
# RACI matrix
| Label | Value | 
| ------------- | ------------- |
| Responsible | @jsartisan, @wmdev0808   |
| Accountable | @somangshu  |
| Consulted | @areyabhishek, @momcilo-appsmith   |
| Informed |  @Nikhil-Nandagopal |

# Objective

We need a checkbox group to allow users to easily configure multiple checkboxes together

# Success Metrics

| Goal  | Metric |
| ------------- | ------------- |
| Decrease the time needed to build a collection of checkbox operating as a unit  | Time that user spends creating a group of checkbox |

# Requirements

| Requirement   | User Story | Importance | Notes |
| ------------- | ------------- | ------------- | ------------- |
| Add checkbox | Ability to add new checkbox & drag to reorder the sequence | High  | - |
| Display as a row | Ability to change the view of checkbox group as a row or column; Boolean/Toggle value | High  | - |
| All properties of checkbox | Inherit all the properties of checkbox widget for each row | High  | - |
| Internal property - `selectedValues` | API reference to get all selected values from the checkbox group  | High  | - |
| Action - `onCheckChanged` | API to trigger an action on check/uncheck of a checkbox inside the group | High  | - |

# Developer Handoff Document in Figma:

https://www.figma.com/file/u0IE1Vqr2MKcQcQKCpHO4O/Checkbox-Group-Widget?node-id=2%3A3

### PR DESCRIPTION
**Pull request by wmdev0808**
## Description

- Build the first MVP of the widget
Fixes #3344 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

- Test A
- Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/3344-checkbox-group-widget 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.76 **(-0.04)** | 36.82 **(0)** | 33.04 **(-0.03)** | 55.33 **(-0.04)**
 :sparkles: :new: | **app/client/src/components/designSystems/appsmith/CheckboxGroupComponent.tsx** | **41.18** | **23.53** | **0** | **50**
 :green_circle: | app/client/src/icons/WidgetIcons.tsx | 95.16 **(0.16)** | 100 **(0)** | 90 **(0.34)** | 95.16 **(0.16)**
 :red_circle: | app/client/src/utils/WidgetRegistry.tsx | 75.93 **(-0.26)** | 100 **(0)** | 27.78 **(-0.79)** | 75.93 **(-0.26)**
 :sparkles: :new: | **app/client/src/widgets/CheckboxGroupWidget.tsx** | **35.21** | **0** | **26.32** | **36.36**
 :red_circle: | app/client/src/workers/validations.ts | 76.34 **(-2.89)** | 69.5 **(-0.45)** | 72 **(-8)** | 78.37 **(-3.31)**</details>

### CODE REVIEW

**Comment by marks0351** [comment]
@wmdev0808 
Comments after looking at the deploy preview:

- Each checkbox inside a checkbox group should be the height of each row else in most cases the group alignment will not be symmetrical. 
- The required border takes up the width of the widget so when its enabled u can see the widget moving coz of decrease in space. try using box-show instead.
- Also with the default dimensions of the dropped widget when required is enabled the widget looks weird.

One weird bug:
- Open prop pane
- Remove value for one of the options(Lemon) keep it empty.
- Then remove value for another(Orange)
- Now add a value for Orange you will see a new entry for Orange. 
This seems to happen with radio group as well.
cc: @somangshu

File: app/client/src/utils/autocomplete/EntityDefinitions.ts [hunk 1]
```diff
@@ -268,6 +268,16 @@ export const entityDefinitions = {
     isVisible: isVisible,
     label: "string",
   },
+  CHECKBOX_GROUP_WIDGET: {
+    "!doc":
+      "Checkbox group widget allows users to easily configure multiple checkboxes together.",
+    "!url": "https://docs.appsmith.com/widget-reference/checkbox-group",
```
Comment (marks0351) [inline app/client/src/utils/autocomplete/EntityDefinitions.ts:274]: Is there a appsmith docs PR for this?

File: app/client/src/mockResponses/WidgetConfigResponse.tsx [hunk 2]
```diff
@@ -1198,6 +1198,20 @@ const WidgetConfigResponse: WidgetConfigReducerState = {
...
     },
+    [WidgetTypes.CHECKBOX_GROUP_WIDGET]: {
+      rows: 2 * GRID_DENSITY_MIGRATION_V1,
+      columns: 1.5 * GRID_DENSITY_MIGRATION_V1,
+      options: [
+        { label: "Apple", value: "apple" },
+        { label: "Orange", value: "orange" },
+        { label: "Lemon", value: "lemon" },
+      ],
+      isDisabled: false,
```
Comment (marks0351) [inline app/client/src/mockResponses/WidgetConfigResponse.tsx:1209]: should we have a default inline value as well?

File: app/client/src/widgets/CheckboxGroupWidget.tsx [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            },
+          },
+        ],
+      },
+      {
+        sectionName: "Actions",
+        children: [
+          {
+            helpText: "Triggers an action when the check state is changed",
+            propertyName: "onCheckChanged",
```
Comment (marks0351) [inline app/client/src/widgets/CheckboxGroupWidget.tsx:130]: can u change this to onSelectionChange

**Review by marks0351 — CHANGES_REQUESTED** [review]

**Comment by marks0351** [comment]
@wmdev0808 Can u also add jest and cypress tests for this widget.

**Comment by YogeshJayaseelan** [comment]
@wmdev0808 @jsartisan Tested this feature and came up with the below issues 

1. Option label and values cannot be duplicated 

Screenshot 2021-07-29 at 1 08 13 PM

2. When you decrease the size of the widget the checkbox value should not go out of the widget 

Screenshot 2021-07-29 at 1 08 36 PM

3. Executing an api and creating map function in the checkbox widget - option JS field and clicking on options the value disappears and the widget is shown empty 

Note: You can use any api and based on the output map functions needs to be created 

API used - https://reqres.in/api/users?page=2 

MapFunction - {{ Api1.data.data.map((row) => {
      return { label: row.first_name, value: row.first_name }
   })
}}

https://loom.com/share/5d85272fead043f8baf9cd8441e181d1 


4. When the default option is deleted and bind-ed with text widget {{checkbox1.Selectedvalue}} the deleted option value is displayed in the text widget 

https://loom.com/share/181798e515ea4321aae8d323ed899792

5. By default user is unable to check and uncheck checkbox and it is observed that only when the default value selected value is passed user is able to check/uncheck 

https://loom.com/share/c2391558567941b08e9f8ad0deb86af9

6. The Default selected value design is not matching the figma design 

Screenshot 2021-07-29 at 2 28 33 PM

7. When the incorrect JS option map function is passed the checkbox widget crashes 

https://www.loom.com/share/db6eb1503aa34f37b3a068bff80b02e5

8. when you added multiple labels and when you delete labels the widget size can be enlarged when added more that default size and reduced whenever an label is removed 

https://www.loom.com/share/f6b9a4229a954169acfc05f17d8b712b

9. Suggestion - For the default selected value - The expected structure can be 'Value or Value of array' similar to select widget 

https://www.loom.com/share/3855037811e141109642dc3d15dff671

**Comment by wmdev0808** [comment]
> @wmdev0808 @jsartisan Tested this feature and came up with the below issues
> 
> 1. Option label and values cannot be duplicated
> 
> Screenshot 2021-07-29 at 1 08 13 PM
> 
> 1. When you decrease the size of the widget the checkbox value should not go out of the widget
> 
> Screenshot 2021-07-29 at 1 08 36 PM
> 
> 1. Executing an api and creating map function in the checkbox widget - option JS field and clicking on options the value disappears and the widget is shown empty
> 
> Note: You can use any api and based on the output map functions needs to be created
> 
> API used - https://reqres.in/api/users?page=2
> 
> MapFunction - {{ Api1.data.data.map((row) => {
> return { label: row.first_name, value: row.first_name }
> })
> }}
> 
> https://loom.com/share/5d85272fead043f8baf9cd8441e181d1
> 
> 1. When the default option is deleted and bind-ed with text widget {{checkbox1.Selectedvalue}} the deleted option value is displayed in the text widget
> 
> https://loom.com/share/181798e515ea4321aae8d323ed899792
> 
> 1. By default user is unable to check and uncheck checkbox and it is observed that only when the default value selected value is passed user is able to check/uncheck
> 
> https://loom.com/share/c2391558567941b08e9f8ad0deb86af9
> 
> 1. The Default selected value design is not matching the figma design
> 
> Screenshot 2021-07-29 at 2 28 33 PM
> 
> 1. When the incorrect JS option map function is passed the checkbox widget crashes
> 
> https://www.loom.com/share/db6eb1503aa34f37b3a068bff80b02e5
> 
> 1. when you added multiple labels and when you delete labels the widget size can be enlarged when added more that default size and reduced whenever an label is removed
> 
> https://www.loom.com/share/f6b9a4229a954169acfc05f17d8b712b
> 
> 1. Suggestion - For the default selected value - The expected structure can be 'Value or Value of array' similar to select widget
> 
> https://www.loom.com/share/3855037811e141109642dc3d15dff671

@

> @wmdev0808 @jsartisan Tested this feature and came up with the below issues
> 
> 1. Option label and values cannot be duplicated
> 
> Screenshot 2021-07-29 at 1 08 13 PM
> 
> 1. When you decrease the size of the widget the checkbox value should not go out of the widget
> 
> Screenshot 2021-07-29 at 1 08 36 PM
> 
> 1. Executing an api and creating map function in the checkbox widget - option JS field and clicking on options the value disappears and the widget is shown empty
> 
> Note: You can use any api and based on the output map functions needs to be created
> 
> API used - https://reqres.in/api/users?page=2
> 
> MapFunction - {{ Api1.data.data.map((row) => {
> return { label: row.first_name, value: row.first_name }
> })
> }}
> 
> https://loom.com/share/5d85272fead043f8baf9cd8441e181d1
> 
> 1. When the default option is deleted and bind-ed with text widget {{checkbox1.Selectedvalue}} the deleted option value is displayed in the text widget
> 
> https://loom.com/share/181798e515ea4321aae8d323ed899792
> 
> 1. By default user is unable to check and uncheck checkbox and it is observed that only when the default value selected value is passed user is able to check/uncheck
> 
> https://loom.com/share/c2391558567941b08e9f8ad0deb86af9
> 
> 1. The Default selected value design is not matching the figma design
> 
> Screenshot 2021-07-29 at 2 28 33 PM
> 
> 1. When the incorrect JS option map function is passed the checkbox widget crashes
> 
> https://www.loom.com/share/db6eb1503aa34f37b3a068bff80b02e5
> 
> 1. when you added multiple labels and when you delete labels the widget size can be enlarged when added more that default size and reduced whenever an label is removed
> 
> https://www.loom.com/share/f6b9a4229a954169acfc05f17d8b712b
> 
> 1. Suggestion - For the default selected value - The expected structure can be 'Value or Value of array' similar to select widget
> 
> https://www.loom.com/share/3855037811e141109642dc3d15dff671

File: app/client/src/mockResponses/WidgetConfigResponse.tsx (hunk 2 above)
Comment (wmdev0808) [reply app/client/src/mockResponses/WidgetConfigResponse.tsx:1209]: No

File: app/client/src/utils/autocomplete/EntityDefinitions.ts (hunk 1 above)
Comment (wmdev0808) [reply app/client/src/utils/autocomplete/EntityDefinitions.ts:299]: I will make it soon

File: app/client/src/widgets/CheckboxGroupWidget.tsx (hunk 3 above)
Comment (wmdev0808) [reply app/client/src/widgets/CheckboxGroupWidget.tsx]: done

**Comment by wmdev0808** [comment]
> @wmdev0808 @jsartisan Tested this feature and came up with the below issues
> 
> 1. Option label and values cannot be duplicated
> 
> Screenshot 2021-07-29 at 1 08 13 PM
> 
> 1. When you decrease the size of the widget the checkbox value should not go out of the widget
> 
> Screenshot 2021-07-29 at 1 08 36 PM
> 
> 1. ~Executing an api and creating map function in the checkbox widget - option JS field and clicking on options the value disappears and the widget is shown empty~ Issue is not related
> 
> Note: You can use any api and based on the output map functions needs to be created
> 
> API used - https://reqres.in/api/users?page=2
> 
> MapFunction - {{ Api1.data.data.map((row) => {
> return { label: row.first_name, value: row.first_name }
> })
> }}
> 
> https://loom.com/share/5d85272fead043f8baf9cd8441e181d1
> 
> 1. When the default option is deleted and bind-ed with text widget {{checkbox1.Selectedvalue}} the deleted option value is displayed in the text widget - this seems like an issue with the evaluation
> 
> https://loom.com/share/181798e515ea4321aae8d323ed899792
> 
> 1. By default user is unable to check and uncheck checkbox and it is observed that only when the default value selected value is passed user is able to check/uncheck
> 
> https://loom.com/share/c2391558567941b08e9f8ad0deb86af9
> 
> 1. The Default selected value design is not matching the figma design
> 
> Screenshot 2021-07-29 at 2 28 33 PM
> 
> 1. When the incorrect JS option map function is passed the checkbox widget crashes
> 
> https://www.loom.com/share/db6eb1503aa34f37b3a068bff80b02e5
> 
> 1. when you added multiple labels and when you delete labels the widget size can be enlarged when added more that default size and reduced whenever an label is removed
> 
> https://www.loom.com/share/f6b9a4229a954169acfc05f17d8b712b
> 
> 1. Suggestion - For the default selected value - The expected structure can be 'Value or Value of array' similar to select widget
> 
> https://www.loom.com/share/3855037811e141109642dc3d15dff671

File: app/client/src/utils/autocomplete/EntityDefinitions.ts (hunk 1 above)
Comment (wmdev0808) [reply app/client/src/utils/autocomplete/EntityDefinitions.ts:299]: I've made a PR for docs

**Comment by wmdev0808** [comment]
> @wmdev0808 Can u also add jest and cypress tests for this widget.

added cypress test

**Review by marks0351 — APPROVED** [review]

**Comment by somangshu** [comment]
@wmdev0808 please check, `CheckboxGroup_spec.js` failed

**Review by somangshu — DISMISSED** [review]
Approving to run tests

**Comment by YogeshJayaseelan** [comment]
@wmdev0808 @somangshu Tested the fix and please find the status below,

**Issues Fixed:**
1,2,4,5,9

**Issues Reopened:**

3 - This is still happening - need confirmation that this is related to checkbox grouping widget. If not we can raise an separate issue 

6. Need confirmation here - Still the Default value doesn't match figma design 

7. This is still happening - But this time the app itself crashes by passing invalid Map function. 

https://www.loom.com/share/02363befef634d658e8626623c35a555

**Steps:**
**7.a)** Copy paste the following map function directly into the option JS field and observe- 
{{ Api1.data.data.map((row) => {
return { label: row.first_name, value: row.first_name }
})
}}
**7.b)** Or execute the following API - https://reqres.in/api/users?page=2 and pass the following valid map function 
{{ Api1.data.data.map((row) => {
return { label: row.first_name, value: row.first_name }
})
}} then modify the map function to invalid 

8) This is fixed - When user adds more options the widget has scroll options. But when the options are deleted the widget size still remains bigger 

https://www.loom.com/share/9f792acfd00c4e48afa23635731e2c6a

**New issue:**
10. 'The value does not evaluate to type string' error occurs when the checkbox selected value is bind-ed to the text widget 
https://loom.com/share/bbbb56bb4cf24629b2aab6e9da7cdd97

**Comment by somangshu** [comment]
@YogeshJayaseelan I have updated your comment above. It was causing confusions. Let me know if you think something is not right

**Comment by YogeshJayaseelan** [comment]
@wmdev0808 @somangshu Retested the fix and please find the status below,

**Issues Fixed :**
2,4,5,9 
3 - Will raise an separate issue 
6 - Expected behaviour 
8- Expected behaviour 
10 - Expected behaviour

**Issues reopened:**

1 - The option and values are non-case sensitive - If the existing option is 'Lemon' and its value is 'lemon' user is able to enter options and value as 'lemon' and 'Lemon' 

https://loom.com/share/db25aa3fd7874a3d872c3c717348a1b9 

7 - When the invalid map function is called - Now 'Reference not found' error is displayed. But still the app crashes when the user removes the values in the options field after getting (Referenced not found or illegal argument errors)

https://loom.com/share/e85769a1e3c94f97b9e6d3037c7d84f8

7.a) Copy paste the following map function directly into the option JS field and observe- {{ Api1.data.data.map((row) => { return { label: row.first_name, value: row.first_name } }) }} - Reference not found error will occur and remove the entire value from option field and observe
or 
7.b) Without any existing api - pass the following value inside option field - {{api1.data}} - Reference not found error wil occur and remove the entire value from option field and observe
or
7.c)execute the following API - https://reqres.in/api/users?page=2 and pass the following valid map function
{{ Api1.data.data.map((row) => { return { label: row.first_name, value: row.first_name } }) }} then modify the map function to invalid - Illegal argument error will occur and removed the value from the option field and observe

**Comment by somangshu** [comment]
@YogeshJayaseelan please retest the changes here

**Review by somangshu — DISMISSED** [review]
Approving to run tests

**Comment by YogeshJayaseelan** [comment]
@wmdev0808 Verified the bug fixes and they are working as expected.

**Comment by somangshu** [comment]
@wmdev0808 the `CheckboxGroup_spec.js` is still failing, Please look into this

**Review by somangshu — APPROVED** [review]
Approving to run tests

**Comment by somangshu** [comment]
@marks0351 all done here, Can you give one sign-off on the review

**Review by marks0351 — APPROVED** [review]

**Comment by areyabhishek** [comment]
@wmdev0808 Some comments:
1. The design of the checkbox doesn't match the Figma designs. 
Screen Shot 2021-08-21 at 12 39 12 AM
2. The checkbox group isn't centered when there's just one option
Screen Shot 2021-08-21 at 12 54 49 AM
3. The minimum gap between items when resized should be 15px like in the Figma files. When you resize the boundaries to smaller than this, then let the items overflow like in the image below for radio buttons.
Screen Shot 2021-08-21 at 1 06 15 AM

@somangshu @wmdev0808 @jsartisan What does the inline option do? The checkbox group moves around when it's resized and it doesn't have a minimum gap.

**Comment by somangshu** [comment]
@areyabhishek
3. The min gap between the items are the min row height in which one item fits, Unlike Radio group we dont let the widget content overflow, rather add a scroll. 

The inline option give a flex wrap like behaviour, which could be a default. 

We will look into the rest.

**Pull request merged**